File: Python/215. Kth Largest Element in an Array(again.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        heapSize = len(nums)
        self.Heapify(nums, heapSize)
        for i in range(1, k):
            index = len(nums) - i
            nums[0] = nums[index]
            heapSize -= 1
            self.maxHeapify(nums, 0, heapSize)
        return nums[0]
    
    def Heapify(self, nums, heapSize):
        i = (int)(heapSize / 2)
        while i >= 0:
            self.maxHeapify(nums, i, heapSize)
            i -= 1

    def maxHeapify(self, nums, base, heapSize):
        largest = base
        left = 2 * base + 1
        right = 2 * base + 2
        if left < heapSize and nums[left] > nums[largest]:
            largest = left
        if right < heapSize and nums[right] > nums[largest]:
            largest = right
        if base != largest:
            nums[base], nums[largest] = nums[largest], nums[base]
            self.maxHeapify(nums, largest, heapSize)

File: Java/215. Kth Largest Element in an Array(again.java
package Java;

// 手写大根堆

class Solution {
    public int findKthLargest(int[] nums, int k) {
        int heapSize = nums.length;
        Heapify(nums, heapSize);
        for (int i = 1; i < k; i++) {
            int index = nums.length - i;
            System.out.println(nums[0]);
            swap(nums, 0, index);
            heapSize--;
            maxHeapify(nums, 0, heapSize);
        }
        return nums[0];
    } 

    public void Heapify(int[] nums, int heapSize) {
        int max = heapSize / 2;
        for (int i = max; i >= 0; i--) {
            maxHeapify(nums, i, heapSize);
        }
    }

    public void maxHeapify(int[] nums, int base, int heapSize) {
        int largest = base, left = 2 * base + 1, right = 2 * base + 2;
        if (left < heapSize && nums[left] > nums[largest]) largest = left;
        if (right < heapSize && nums[right] > nums[largest]) largest = right;
        if (largest != base) {
            swap(nums, base, largest);
            maxHeapify(nums, largest, heapSize);
        }
    }

    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}